yarn-project/prover-client/src/orchestrator-2/circuits/public-kernel-non-tail.ts

```diff
@@ -0,0 +1,71 @@
+import {
+  AvmProofAndVerificationKey,
+  NestedRecursiveProofAndVK,
+  ProvingRequestType,
+  PublicInputsAndRecursiveProof,
+  PublicKernelNonTailRequest,
+  TubeProofAndVK,
+} from '@aztec/circuit-types';
+import { Proof, PublicKernelCircuitPrivateInputs, PublicKernelCircuitPublicInputs } from '@aztec/circuits.js';
+import { memoize } from '@aztec/foundation/decorators';
+import { createDebugLogger } from '@aztec/foundation/log';
+import { promiseWithResolvers } from '@aztec/foundation/promise';
+
+import { Circuit, OrchestratorContext } from '../types.js';
+
+type PublicKernelPreviousProof = NestedRecursiveProofAndVK | TubeProofAndVK;
+
+export class PublicKernelNonTailCircuit implements Circuit<typeof ProvingRequestType.PUBLIC_KERNEL_NON_TAIL> {
+  private logger = createDebugLogger('aztec:prover-client:public-kernel-non-tail');
+
+  private previousKernelProof = promiseWithResolvers<PublicKernelPreviousProof>();
+  private nestedAvmProof = promiseWithResolvers<AvmProofAndVerificationKey>();
+
+  constructor(public readonly provingRequest: PublicKernelNonTailRequest, private context: OrchestratorContext) {}
+
+  public setPreviousKernelProof(proof: PublicKernelPreviousProof) {
+    this.previousKernelProof.resolve(proof);
+  }
+
+  public setNestedAvmProof(proof: AvmProofAndVerificationKey) {
+    this.nestedAvmProof.resolve(proof);
+  }
+
+  @memoize
+  private async getProvingInputs(): Promise<PublicKernelCircuitPrivateInputs> {
+    const previousKernelProof = await this.previousKernelProof.promise;
+    const avmProof = await this.nestedAvmProof.promise;
+
+    // TODO: The kernel does not require the AVM VK, we're just getting the proof. Is that correct?
+    // TODO(#7369): How should we properly set up vkPath, vkIndex, and clientIvc for previous kernel?
```

Comment:
I'm not sure I follow. All of the public kernels are recursively verified at the moment.